script/scripts/MigrateAccounting.s.sol

```diff
@@ -0,0 +1,170 @@
+        assertEq(_snapshot.tokenSeller, address(_pool.tokenSeller()));
+        assertEq(_snapshot.kycManager, address(ZkBobAccounting(address(_pool.accounting())).kycProvidersManager()));
+        assertEq(accountingOwner, ZkBobAccounting(address(_pool.accounting())).owner());
+
+        checkSlot0(uint256(_snapshot.slot0), ZkBobAccounting(address(_pool.accounting())));
+        checkSlot1(uint256(_snapshot.slot1), ZkBobAccounting(address(_pool.accounting())));
+    }
+
+    function checkSlot0(uint256 slot0, ZkBobAccounting accounting) internal {
+        (
+            uint56 maxWeeklyAvgTvl,
+            uint32 maxWeeklyTxCount,
+            uint24 tailSlot,
+            uint24 headSlot,
+            uint88 cumTvl,
+            uint32 txCount
+        ) = accounting.slot0();
+        uint24 curSlot = uint24(block.timestamp / 1 hours);
+
+        assertEq(uint56(slot0), maxWeeklyAvgTvl);
+        assertEq(uint32(slot0 >> 56), maxWeeklyTxCount);
+        assertEq(curSlot, tailSlot);
+        assertEq(curSlot, headSlot);
+        assertEq(uint88(slot0 >> (56 + 32 + 24 + 24)), cumTvl);
+        assertEq(uint32(slot0 >> (56 + 32 + 24 + 24 + 88)), txCount);
+    }
+
+    function checkSlot1(uint256 slot1, ZkBobAccounting accounting) internal {
+        (uint72 tvl) = accounting.slot1();
+        assertEq(uint72(slot1), tvl);
+    }
+}
+
+/**
+ * @dev Don't forget to set ZkBobPool.TOKEN_NUMERATOR to 1000 for USDC pools.
+ */
+contract MigrateAccounting is Script, UpgradeTest {
+    function run() external {
+        ZkBobPoolUSDC pool = ZkBobPoolUSDC(address(zkBobPool));
+        PoolSnapshot memory snapshot = makeSnapshot(pool);
+
+        vm.startBroadcast();
+
+        // 1. Deploy new ZkBobPoolUSDC implementation
+        ZkBobPoolUSDC newImpl = new ZkBobPoolUSDC(
+            pool.pool_id(),
+            pool.token(),
+            pool.transfer_verifier(),
+            pool.tree_verifier(),
+            pool.batch_deposit_verifier(),
+            address(pool.direct_deposit_queue())
+        );
+
+        // 2. Upgrade proxy to new implementation
+        EIP1967Proxy(payable(address(pool))).upgradeTo(address(newImpl));
+
+        migrateAccounting(address(pool), address(snapshot.kycManager));
+
+        vm.stopBroadcast();
+
+        postCheck(ZkBobPoolUSDC(address(pool)), snapshot);
+    }
+
+    // TODO: Check limits
+    function migrateAccounting(address _pool, address _kycManager) internal {
+        // 3. Deploy new ZkBobAccounting implementation
+        ZkBobAccounting accounting = new ZkBobAccounting(address(_pool), 1_000_000_000);
+
+        bytes memory dump = ZkBobPool(_pool).extsload(bytes32(uint256(1)), 2);
+        uint32 txCount = uint32(_load(dump, 0, 4));
+        uint88 cumTvl = uint88(_load(dump, 4, 11));
+        uint32 maxWeeklyTxCount = uint32(_load(dump, 21, 4));
+        uint56 maxWeeklyAvgTvl = uint56(_load(dump, 25, 7));
+        uint72 tvl = uint72(_load(dump, 55, 9));
```

Comment:
The problem with such approach is that we won't be able to build a static Safe tx with dynamic arguments. That's why prior script had a Migrator contract Safe delegatecall-ed to. In the current setting the owner is also a Timelock, which doesn't support delegatecall, so I propose to organize the script in the following way:
1. Deploy new ZkBobPoolUSDC and new ZkBobAccounting outside of Safe proposal.
2. Deploy a special one-time Migrator contract (similar to how the previous script does), also outside of Safe proposal.
3. Inside a timelock proposal, do the following batch of calls:
  a. Update pool implementation.
  b. Transfer proxy ownership to the Migrator contract.
  c. Call the migrator contract, which does everything in `migrateAccounting` function and then transfers the proxy ownership back to the timelock.

The Safe proposal for the Timelock contract then would be just a `scheduleBatch` with 3 calls.

Reply:
Thank you, I overlooked that. I've separated the actions into three scripts as you suggested. The only difference is that it seems we also need to transfer ownership first to the timelock, then to the migrator, and finally to the specified owner in the migrator contract.